src/Microsoft.ML.Data/Transforms/RowToRowTransformerBase.cs

```diff
@@ -0,0 +1,107 @@
+        }
+
+        protected abstract IRowMapper MakeRowMapper(Schema schema);
+
+        public Schema GetOutputSchema(Schema inputSchema)
+        {
+            Host.CheckValue(inputSchema, nameof(inputSchema));
+            var mapper = MakeRowMapper(inputSchema);
+            return RowToRowMapperTransform.GetOutputSchema(inputSchema, mapper);
+        }
+
+        public IDataView Transform(IDataView input) => MakeDataTransform(input);
+
+        protected RowToRowMapperTransform MakeDataTransform(IDataView input)
+        {
+            Host.CheckValue(input, nameof(input));
+            return new RowToRowMapperTransform(Host, input, MakeRowMapper(input.Schema), MakeRowMapper);
+        }
+
+        protected abstract class MapperBase : IRowMapper
+        {
+            protected readonly IHost Host;
+            protected readonly Schema InputSchema;
+            protected readonly Schema.Column[] OutputColumns;
+
+            protected MapperBase(IHost host, Schema inputSchema)
+            {
+                Contracts.CheckValue(host, nameof(host));
+                Contracts.CheckValue(inputSchema, nameof(inputSchema));
+                Host = host;
+                InputSchema = inputSchema;
+                OutputColumns = GetOutputColumnsCore();
+            }
+
+            protected abstract Schema.Column[] GetOutputColumnsCore();
+
+            public Schema.Column[] GetOutputColumns() => OutputColumns;
+
+            public Delegate[] CreateGetters(IRow input, Func<int, bool> activeOutput, out Action disposer)
+            {
+                // REVIEW: it used to be that the mapper's input schema in the constructor was required to be reference-equal to the schema
+                // of the input row.
+                // It still has to be the same schema, but because we may make a transition from lazy to eager schema, the reference-equality
+                // is no longer always possible. So, we relax the assert as below.
+                if (input.Schema is Schema s)
+                    Contracts.Assert(s == InputSchema);
+                var result = new Delegate[OutputColumns.Length];
+                var disposers = new Action[OutputColumns.Length];
+                for (int i = 0; i < OutputColumns.Length; i++)
+                {
+                    if (!activeOutput(i))
+                        continue;
+                    result[i] = MakeGetter(input, i, activeOutput, out disposers[i]);
+                }
+                if (disposers.Any(x => x != null))
+                {
+                    disposer = () =>
+                    {
+                        foreach (var act in disposers)
+                            act();
+                    };
+                }
+                else
+                    disposer = null;
+                return result;
+            }
+
+            protected abstract Delegate MakeGetter(IRow input, int iinfo, Func<int, bool> activeOutput, out Action disposer);
```

Comment:
>Func<int, bool> activeOutput [](start = 74, length = 28)

follow up question on this -- why add this Func<> delegate to MakeGetter ?

Reply:
In this PR mostly for TensorFlow, (but Onnx transform takes same approach)
They have that thing:
activeOutputColNames for caching purpose and only way for me to acquire it is by passing that activeOutput function.
It's not used in other transforms.

Reply:
thanks for the info. i will look into the TfTransform usage. 

---
In reply to: [235108307](https://github.com/dotnet/machinelearning/pull/1655#discussion_r235108307) [](ancestors = 235108307)